drivers/i2c/dw_i2c_master.c

```diff
@@ -288,14 +285,20 @@ uint8_t i2c_hw_read_byte(void)
 		dw_i2c_pri->state = DW_I2C_DRIVER_INVALID;
 		return 0xFF; // XXX Invalid value
 	}
+
 	/* When IC_RX_TL = 0 (set in _init()), RX_FULL will be reported as
 	 * soon as one byte is received. */
-	// TODO: Detect timeout
+	check_cnt = 0;
 	do {
 		val = __raw_readl(base + IC_INTR_STAT);
-	} while (!(val & IC_RX_FULL));
-	val = __raw_readl(base + IC_DATA_CMD);
-	return (uint8_t)val;
+		if (val & IC_RX_FULL) {
+			val = __raw_readl(base + IC_DATA_CMD);
+			return (uint8_t)val;
+		}
+		check_cnt++;
+		udelay(200);
+	} while (check_cnt <= 10); /* Wait max 2ms for each byte. */
+	return 0xFF; // XXX Invalid value
```

Comment:
Should use tick_get_counter() and time_before/after() to achieve busy waiting code.